include/pmacc/math/vector/Vector.hpp

```diff
@@ -700,36 +724,112 @@ namespace pmacc
             return result;
         }
 
+        /** Give the linear index of a N-dimensional index within an N-dimensional index space.
+         *
+         * @tparam T_IntegralType vector data type (must be an integral type)
+         * @tparam T_dim dimension of the vector, should be >= 2
+         * @param size N-dimensional size of the index space (N can be one dimension less compared to idx)
+         * @param idx N-dimensional index within the index space
+         *            @attention behaviour is undefined for negative index
+         *            @attention if idx is outside of size the result will be outside of the the index domain too
+         * @return linear index within the index domain
+         *
+         * @{
+         */
         template<
-            typename T_Type,
+            typename T_IntegralType,
             typename T_Accessor,
             typename T_Navigator,
             typename T_Storage,
             typename T_OtherAccessor,
             typename T_OtherNavigator,
-            typename T_OtherStorage>
-        HDINLINE T_Type linearize(
-            const Vector<T_Type, 1, T_Accessor, T_Navigator, T_Storage>& size,
-            const Vector<T_Type, 2, T_OtherAccessor, T_OtherNavigator, T_OtherStorage>& pos)
+            typename T_OtherStorage,
+            uint32_t T_dim,
+            typename = std::enable_if_t<std::is_integral_v<T_IntegralType> && T_dim >= DIM2>>
+        HDINLINE T_IntegralType linearize(
+            const Vector<T_IntegralType, T_dim - 1u, T_Accessor, T_Navigator, T_Storage>& size,
+            const Vector<T_IntegralType, T_dim, T_OtherAccessor, T_OtherNavigator, T_OtherStorage>& idx)
         {
-            return pos.y() * size.x() + pos.x();
+            auto pitchExtents = size;
+            for(uint32_t d = 1u; d < T_dim - 1u; ++d)
+                pitchExtents[d] *= pitchExtents[d - 1u];
+
+            T_IntegralType linearIdx = idx[0];
+            for(uint32_t d = 1u; d < T_dim; ++d)
+                linearIdx += pitchExtents[d - 1u] * idx[d];
+            return linearIdx;
         }
```

Comment:
T_IntegralType linearIdx{idx[T_dim - 1]};
            for(int i = T_dim - 2; i >= 0; i--)
                linearIdx = linearIdx * pitchExtents[i] + idx[i];
            ```suggestion
            auto pitchExtents = size;
            for(uint32_t d = 1u; d < T_dim - 1u; ++d)
                pitchExtents[d] *= pitchExtents[d - 1u];

            T_IntegralType linearIdx = idx[0];
            for(uint32_t d = 1u; d < T_dim; ++d)
                linearIdx += pitchExtents[d - 1u] * idx[d];
            ```